### PR TITLE
Feature/search filter

### PR DESCRIPTION
**Pull request by ghost**
## Summary

* Create filter header
* Create filter screen
* Get filtered books using redux-recompose
* Implemente a basic debounce hook

#### Galaxy Nexus
image

### Pixe 2
image

#### Pixel 3XL
image

#### iPhone SE
image

#### iPhone 8
image

#### iPhone 12 Pro Max
image

## Trello Card

[Trello Card](https://trello.com/c/yINdEP1F/31-n-112-filtro-de-b%C3%BAsqueda)

### CODE REVIEW

File: wbooks/src/app/components/AppNavigator/index.tsx [hunk 1]
```diff
@@ -19,10 +22,27 @@ function AppNavigator() {
         headerBackground: HeaderBackground,
         headerTitleStyle: styles.headerTitle,
         headerBackImage: HeaderBackButton,
-        headerBackTitleVisible: false
+        headerBackTitleVisible: false,
+        headerRight: HeaderRightButton
```
Comment (sfernandez11) [inline wbooks/src/app/components/AppNavigator/index.tsx]: This must be only on the Library screen, not in the others

File: wbooks/src/app/components/AppNavigator/index.tsx [hunk 2]
```diff
@@ -19,10 +22,27 @@ function AppNavigator() {
...
       }}>
       <LibraryStack.Screen name={ROUTES.library} component={LibraryScreen} />
-      <LibraryStack.Screen name={ROUTES.bookDetail} component={BookDetailScreen} />
+      <LibraryStack.Screen
+        name={ROUTES.bookDetail}
+        component={BookDetailScreen}
+        options={{
+          headerRight: () => <></>
+        }}
```
Comment (sfernandez11) [inline wbooks/src/app/components/AppNavigator/index.tsx]: Do it right and you don't need this

File: wbooks/src/app/components/AppNavigator/index.tsx [hunk 3]
```diff
@@ -19,10 +22,27 @@ function AppNavigator() {
...
+      />
+      <LibraryStack.Screen
+        name={ROUTES.filter}
+        component={FilterScreen}
+        options={{
+          headerTitle: FilterInput,
+          headerRight: () => null,
```
Comment (sfernandez11) [inline wbooks/src/app/components/AppNavigator/index.tsx]: Same here

File: wbooks/src/app/components/AppNavigator/index.tsx [hunk 4]
```diff
@@ -19,10 +22,27 @@ function AppNavigator() {
...
+          headerLeft: () => null,
```
Comment (sfernandez11) [inline wbooks/src/app/components/AppNavigator/index.tsx]: Why? How the user can go back if you hide the back arrow?

File: wbooks/src/app/components/FilterInput/index.tsx [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+import closeDisable from '@assets/general/close_disable.png';
+import close from '@assets/general/close.png';
+import BookActions from '@redux/book/actions';
+import useDebounce from '@hooks/useDebounce';
+
+import styles from './styles';
+
+function FilterInput() {
+  const [value, setValue] = useState('');
+  const debounceSearchTerm = useDebounce(value, 300);
```
Comment (sfernandez11) [inline wbooks/src/app/components/FilterInput/index.tsx]: 300 => constant, no magic numbers

File: wbooks/src/app/components/FilterInput/index.tsx [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+  useEffect(() => {
+    if (debounceSearchTerm) {
+      dispatch(BookActions.getFilteredBooks(value));
+    }
+  }, [debounceSearchTerm, dispatch, value]);
+
+  const onHandlerChange = (text: string) => {
+    setValue(text);
+  };
```
Comment (sfernandez11) [inline wbooks/src/app/components/FilterInput/index.tsx]: Unnecessary I think, you can use `setValue` directly

File: wbooks/src/app/components/FilterInput/index.tsx [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+  return (
+    <View style={styles.container}>
+      <View style={styles.subContainer}>
+        <Image style={styles.icon} resizeMode="cover" source={icSearchPlaceholder} />
+        <TextInput
+          onChangeText={onHandlerChange}
+          value={value}
+          style={styles.input}
+          placeholder="search"
+          placeholderTextColor="gray"
```
Comment (sfernandez11) [inline wbooks/src/app/components/FilterInput/index.tsx]: Use colors constants file

File: wbooks/src/app/components/FilterInput/index.tsx [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+        />
+        <TouchableOpacity onPress={onHandlerCloseIconPress} disabled={value === ''}>
+          <Image style={styles.icon} resizeMode="cover" source={value ? close : closeDisable} />
```
Comment (sfernandez11) [inline wbooks/src/app/components/FilterInput/index.tsx]: Change tintColor, not the asset

File: wbooks/src/app/components/FilterInput/styles.ts [hunk 9]
```diff
@@ -0,0 +1,30 @@
+import { StyleSheet, Platform, Dimensions } from 'react-native';
+import { COLORS } from '@constants/colors';
+
+const { width } = Dimensions.get('window');
+
+export default StyleSheet.create({
+  container: {
+    width,
+    paddingHorizontal: 12.5
```
Comment (sfernandez11) [inline wbooks/src/app/components/FilterInput/styles.ts]: Use round numbers

File: wbooks/src/app/components/FilterInput/styles.ts [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+  subContainer: {
+    borderRadius: 14,
+    paddingHorizontal: 9,
+    backgroundColor: COLORS.white,
+    flexDirection: 'row',
+    alignItems: 'center'
+  },
+  icon: {
+    width: 13,
+    height: 13
```
Comment (sfernandez11) [inline wbooks/src/app/components/FilterInput/styles.ts]: Extract equal values as constant

File: wbooks/src/app/components/HeaderRightButton/index.tsx [hunk 11]
```diff
@@ -0,0 +1,21 @@
+import React from 'react';
+import { Image, TouchableOpacity } from 'react-native';
+import { useNavigation } from '@react-navigation/native';
+import icSearch from '@assets/general/ic_search.png';
+import { ROUTES } from '@constants/routes';
+
+import styles from './styles';
+
+function HeaderRightButton() {
```
Comment (sfernandez11) [inline wbooks/src/app/components/HeaderRightButton/index.tsx]: Use representative names, `HeaderRightButton` => `SearchIcon` or `HeaderSearchIcon` maybe

File: wbooks/src/app/screens/Filter/index.tsx [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+        style={styles.subContainer}
+        showsVerticalScrollIndicator={false}
+        data={books}
+        keyExtractor={getKeyExtractor}
+        renderItem={renderItem}
+        ListEmptyComponent={renderEmptyComponent}
+        contentContainerStyle={books.length === 0 && styles.emptyContainer}
+      />
+    </SafeAreaView>
+  );
```
Comment (sfernandez11) [inline wbooks/src/app/screens/Filter/index.tsx]: Don't repeat code for the `BooksList`, create a general component instead for the Filter and the Library screens

File: wbooks/src/app/screens/Filter/index.tsx [hunk 13]
```diff
@@ -0,0 +1,39 @@
+import React from 'react';
+import { SafeAreaView, FlatList, ListRenderItem } from 'react-native';
+import { useSelector } from 'react-redux';
+import { FilteredBooks, BookState, Book } from '@interfaces/Book';
+import BookComponent from '@components/Book';
+
+import EmptyList from './components/EmptyList';
+import styles from './styles';
+
+function Library() {
```
Comment (sfernandez11) [inline wbooks/src/app/screens/Filter/index.tsx]: Bad copy/paste, be careful!

File: wbooks/src/app/screens/Library/index.tsx [hunk 14]
```diff
@@ -3,8 +3,8 @@ import { SafeAreaView, FlatList, ListRenderItem } from 'react-native';
 import { useDispatch, useSelector } from 'react-redux';
 import { Book, BookState } from '@interfaces/Book';
 import BookActions from '@redux/book/actions';
+import BookComponent from '@components/Book';
```
Comment (sfernandez11) [inline wbooks/src/app/screens/Library/index.tsx]: Extract `BooksList`, not `BookComponent`

File: wbooks/src/redux/book/actions.ts [hunk 15]
```diff
@@ -1,15 +1,22 @@
...
     type: actions.GET_BOOKS,
-    target: TARGET,
+    target: 'books',
     service: getBooks
+  }),
+  getFilteredBooks: (query: string) => ({
+    type: actions.GET_FILTERED_BOOKS,
+    target: 'filteredBooks',
+    service: getFilteredBooks,
+    payload: query
```
Comment (sfernandez11) [inline wbooks/src/redux/book/actions.ts:19]: Just as a comment: In the next card you must use `useMemo` and do the filter in the Filter screen, not in the service/backend

**Review by sfernandez11 — CHANGES_REQUESTED** [review]
Great work! Don't add the same asset twice, use `tintColor` instead
image

File: wbooks/src/app/components/AppNavigator/index.tsx (hunk 4 above)
Comment (ghost) [reply wbooks/src/app/components/AppNavigator/index.tsx]: I design the screen equal to the [Marvel](https://marvelapp.com/prototype/b3fe593/screen/71449389/handoff), as you can see, this screen doesn't have back button

File: wbooks/src/app/screens/Library/index.tsx (hunk 14 above)
Comment (ghost) [reply wbooks/src/app/screens/Library/index.tsx]: I extracted BookList, but BookComponent is a common component between Library and Filter, so I think that this extraction is okay, let me know what dow you think.

File: wbooks/src/app/components/AppNavigator/index.tsx (hunk 1 above)
Comment (ghost) [reply wbooks/src/app/components/AppNavigator/index.tsx]: Done!

File: wbooks/src/app/components/AppNavigator/index.tsx (hunk 2 above)
Comment (ghost) [reply wbooks/src/app/components/AppNavigator/index.tsx]: Done!

File: wbooks/src/app/components/AppNavigator/index.tsx (hunk 3 above)
Comment (ghost) [reply wbooks/src/app/components/AppNavigator/index.tsx]: Done!

File: wbooks/src/app/components/FilterInput/index.tsx (hunk 5 above)
Comment (ghost) [reply wbooks/src/app/components/FilterInput/index.tsx]: Done!

File: wbooks/src/app/components/FilterInput/index.tsx (hunk 6 above)
Comment (ghost) [reply wbooks/src/app/components/FilterInput/index.tsx]: Done!

File: wbooks/src/app/components/FilterInput/index.tsx (hunk 7 above)
Comment (ghost) [reply wbooks/src/app/components/FilterInput/index.tsx]: The [Marvel](https://marvelapp.com/prototype/b3fe593/screen/71449389/handoff) doesn't have placeholder, so I removed `placeholder` and `placeholderTextColor` props

File: wbooks/src/app/components/FilterInput/index.tsx (hunk 8 above)
Comment (ghost) [reply wbooks/src/app/components/FilterInput/index.tsx]: Done!

File: wbooks/src/app/components/FilterInput/styles.ts (hunk 9 above)
Comment (ghost) [reply wbooks/src/app/components/FilterInput/styles.ts]: Done!

File: wbooks/src/app/components/FilterInput/styles.ts (hunk 10 above)
Comment (ghost) [reply wbooks/src/app/components/FilterInput/styles.ts]: Done!

File: wbooks/src/app/components/HeaderRightButton/index.tsx (hunk 11 above)
Comment (ghost) [reply wbooks/src/app/components/HeaderRightButton/index.tsx]: Done!

File: wbooks/src/app/screens/Filter/index.tsx (hunk 12 above)
Comment (ghost) [reply wbooks/src/app/screens/Filter/index.tsx]: Done!

File: wbooks/src/app/screens/Filter/index.tsx (hunk 13 above)
Comment (ghost) [reply wbooks/src/app/screens/Filter/index.tsx]: Done!

File: wbooks/src/redux/book/actions.ts (hunk 15 above)
Comment (ghost) [reply wbooks/src/redux/book/actions.ts:19]: Okay, so in the next card I need to delete the `getFilteredBooks` `service`, `reducer` and `action`, right?

File: wbooks/src/app/components/AppNavigator/index.tsx (hunk 4 above)
Comment (ghost) [reply wbooks/src/app/components/AppNavigator/index.tsx]: Done!

File: wbooks/src/app/components/BookList/index.tsx [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+  return (
+    <SafeAreaView style={styles.container}>
+      <FlatList
+        style={styles.subContainer}
+        showsVerticalScrollIndicator={false}
+        data={books}
+        keyExtractor={getKeyExtractor}
+        renderItem={renderItem}
+        ListEmptyComponent={showEmptyComponent ? renderEmptyComponent : null}
+        contentContainerStyle={books && books.length === 0 && styles.emptyContainer}
```
Comment (sfernandez11) [inline wbooks/src/app/components/BookList/index.tsx]: `!books?.length && styles.emptyContainer`

File: wbooks/src/app/components/FilterInput/index.tsx [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+import { View, TextInput, Image, TouchableOpacity } from 'react-native';
+import { useDispatch } from 'react-redux';
+import icSearchPlaceholder from '@assets/general/ic_search_placeholder.png';
+import close from '@assets/general/close.png';
+import BookActions from '@redux/book/actions';
+import useDebounce from '@hooks/useDebounce';
+
+import styles from './styles';
+
+const delay: number = 300;
```
Comment (sfernandez11) [inline wbooks/src/app/components/FilterInput/index.tsx]: 2 things:

* Again constants must be un UPPERCASE
* Unnecessary TS interface

File: wbooks/src/app/components/FilterInput/styles.ts [hunk 18]
```diff
@@ -0,0 +1,41 @@
+import { StyleSheet, Platform, Dimensions } from 'react-native';
+import { COLORS } from '@constants/colors';
+
+const { width } = Dimensions.get('window');
+const searchIconSize = 13;
+const closeIconSize = 17;
```
Comment (sfernandez11) [inline wbooks/src/app/components/FilterInput/styles.ts]: Same, UPPERCASE

**Review by sfernandez11 — COMMENTED** [review]
Last comments...

File: wbooks/src/app/components/FilterInput/index.tsx (hunk 17 above)
Comment (ghost) [reply wbooks/src/app/components/FilterInput/index.tsx]: Done!

File: wbooks/src/app/components/FilterInput/styles.ts (hunk 18 above)
Comment (ghost) [reply wbooks/src/app/components/FilterInput/styles.ts]: Done!

File: wbooks/src/app/components/BookList/index.tsx (hunk 16 above)
Comment (ghost) [reply wbooks/src/app/components/BookList/index.tsx]: Done!

**Review by sfernandez11 — APPROVED** [review]
🥇

**Pull request merged**